internal/patchpkg/patch.go

```diff
@@ -12,56 +12,80 @@ import (
 	"strings"
 )
 
-// glibcPatcher patches ELF binaries to use an alternative version of glibc.
-type glibcPatcher struct {
+// libPatcher patches ELF binaries to use an alternative version of glibc.
+type libPatcher struct {
 	// ld is the absolute path to the new dynamic linker (ld.so).
 	ld string
 
 	// rpath is the new RPATH with the directories containing the new libc
 	// shared objects (libc.so) and other libraries.
 	rpath []string
-}
 
-// newGlibcPatcher creates a new glibcPatcher and verifies that it can find the
-// shared object files in glibc.
-func newGlibcPatcher(glibc *packageFS) (*glibcPatcher, error) {
-	patcher := &glibcPatcher{}
+	// needed are shared libraries to add as dependencies (DT_NEEDED).
+	needed []string
+}
 
+// setGlibc configures the patcher to use the dynamic linker and libc libraries
+// in pkg.
+func (p *libPatcher) setGlibc(pkg *packageFS) error {
 	// Verify that we can find a directory with libc in it.
 	glob := "lib*/libc.so*"
-	matches, _ := fs.Glob(glibc, glob)
+	matches, _ := fs.Glob(pkg, glob)
 	if len(matches) == 0 {
-		return nil, fmt.Errorf("cannot find libc.so file matching %q", glob)
+		return fmt.Errorf("cannot find libc.so file matching %q", glob)
 	}
 	for i := range matches {
 		matches[i] = path.Dir(matches[i])
 	}
 	slices.Sort(matches) // pick the shortest name: lib < lib32 < lib64 < libx32
 
-	lib, err := glibc.OSPath(matches[0])
+	lib, err := pkg.OSPath(matches[0])
 	if err != nil {
-		return nil, err
+		return err
 	}
-	patcher.rpath = append(patcher.rpath, lib)
+	p.rpath = append(p.rpath, lib)
 	slog.Debug("found new libc directory", "path", lib)
 
 	// Verify that we can find the new dynamic linker.
 	glob = "lib*/ld-linux*.so*"
-	matches, _ = fs.Glob(glibc, glob)
+	matches, _ = fs.Glob(pkg, glob)
 	if len(matches) == 0 {
-		return nil, fmt.Errorf("cannot find ld.so file matching %q", glob)
+		return fmt.Errorf("cannot find ld.so file matching %q", glob)
 	}
 	slices.Sort(matches)
-	patcher.ld, err = glibc.OSPath(matches[0])
+	p.ld, err = pkg.OSPath(matches[0])
 	if err != nil {
-		return nil, err
+		return err
 	}
-	slog.Debug("found new dynamic linker", "path", patcher.ld)
+	slog.Debug("found new dynamic linker", "path", p.ld)
+	return nil
+}
 
-	return patcher, nil
+// setGlibc configures the patcher to use the standard C++ and gcc libraries in
+// pkg.
+func (p *libPatcher) setGcc(pkg *packageFS) error {
+	// Verify that we can find a directory with libstdc++.so in it.
+	glob := "lib*/libstdc++.so*"
+	matches, _ := fs.Glob(pkg, glob)
+	if len(matches) == 0 {
+		return fmt.Errorf("cannot find libstdc++.so file matching %q", glob)
+	}
+	for i := range matches {
+		matches[i] = path.Dir(matches[i])
+	}
+	slices.Sort(matches) // pick the shortest name: lib < lib32 < lib64 < libx32
```

Comment:
why this heuristic?

Reply:
Added a comment.

- lib is usually a symlink to the correct arch (e.g., lib -> lib64)
- *.so is usually a symlink to the correct version (e.g., foo.so -> foo.so.2)